apps/election-manager/src/components/ExportFinalResultsModal.tsx

```diff
@@ -0,0 +1,299 @@
+import React, { useContext, useState, useEffect } from 'react'
+import styled from 'styled-components'
+import path from 'path'
+import fileDownload from 'js-file-download'
+
+import AppContext from '../contexts/AppContext'
+import Modal from './Modal'
+import Button from './Button'
+import Prose from './Prose'
+import LinkButton from './LinkButton'
+import Loading from './Loading'
+import TextInput from './TextInput'
+import Text from './Text'
+import { MainChild } from './Main'
+import USBControllerButton from './USBControllerButton'
+import ConverterClient from '../lib/ConverterClient'
+import { UsbDriveStatus, getDevicePath } from '../lib/usbstick'
+import { generateFinalExportDefaultFilename } from '../utils/filenames'
+
+function throwBadStatus(s: never): never {
+  throw new Error(`Bad status: ${s}`)
+}
+
+const USBImage = styled.img`
+  margin-right: auto;
+  margin-left: auto;
+  height: 200px;
+`
+
+const TextInputRow = styled.span`
+  display: flex;
+  align-items: center;
+  margin-top: 0;
+  font-size: 0.8em;
+  & > *:not(:first-child) {
+    margin-left: 5px;
+  }
+`
+
+const FilenameInput = styled(TextInput)`
+  text-align: right;
+`
+
+const LabelText = styled(Text)`
+  margin-bottom: 0;
+  font-weight: 700;
+`
+
+export interface Props {
+  isOpen: boolean
+  onClose: () => void
+}
+
+enum ModalState {
+  ERROR = 'error',
+  SAVING = 'saving',
+  DONE = 'done',
+  INIT = 'init',
+}
+
+const ExportFinalResultsModal: React.FC<Props> = ({
+  isOpen,
+  onClose: onCloseFromProps,
+}) => {
+  const {
+    usbDriveStatus,
+    castVoteRecordFiles,
+    electionDefinition,
+  } = useContext(AppContext)
+
+  const [currentState, setCurrentState] = useState<ModalState>(ModalState.INIT)
+  const [errorMessage, setErrorMessage] = useState('')
+
+  const [filename, setFilename] = useState('')
+
+  const isTestMode = castVoteRecordFiles?.fileMode === 'test'
+
+  useEffect(() => {
+    if (
+      filename === '' &&
+      (usbDriveStatus === UsbDriveStatus.mounted || !window.kiosk)
+    ) {
+      setFilename(generateFinalExportDefaultFilename(isTestMode))
+    }
+  }, [usbDriveStatus, isOpen])
+
+  const onClose = () => {
+    setErrorMessage('')
+    setCurrentState(ModalState.INIT)
+    setFilename('')
+    onCloseFromProps()
+  }
+
+  const handleFilenameInputChange = (
+    event: React.ChangeEvent<HTMLInputElement>
+  ) => {
+    let newValue = event.currentTarget.value
+    // Only allow characters that are valid in filenames.
+    newValue = newValue.replace(/[^0-9a-zA-Z_-]/gi, '')
```

Comment:
Maybe replace any run of invalid characters with a dash? Maybe also trim dashes off the ends?
```suggestion
    newValue = newValue.replace(/[^0-9a-zA-Z_-]+/gi, '-').replace(/^-+|-+$/g, '')
```